jwt/api_jws.py

```diff
@@ -86,6 +86,10 @@ def encode(
         if algorithm is None:
             algorithm = "none"
 
+        # Prefer headers["alg"] if present to algorithm parameter.
+        if headers and "alg" in headers and headers["alg"]:
+            algorithm = headers["alg"]
```

Comment:
Would type definitions for `algorithm` need updating?

Reply:
If `header["alg"]` is always preferred to `algorithm`, then no. It can remain a `str`.

But lines 86-87 (`if algorithm is None: ...`) are unneeded. I'd remove that.

Reply:
Since `algorithm` parameter is practically optional and for keeping backward-compat, I don't think its type definition should be updated at least for now.

> But lines 86-87 (if algorithm is None: ...) are unneeded. I'd remove that.

As the current spec, `algorithm=None` is interpreted to `{"alg": "none"}` and related tests exist. So it should not be changed for keeping backward-compat for now. It needs another PR if needed.

Reply:
Shouldn't it then already be of type `Optional[str]`? 🤔 Passing `algorithm=None` fails the typecheck with the current annotation. Keeping it annotated as a `str` only makes sense if that option will (eventually) be removed.

Reply:
> Shouldn't it then already be of type Optional[str]? Passing algorithm=None fails the typecheck with the current annotation.

I agree with you and I have recognized this problem. I doubt this fix need to be done in this PR but I'll fix it.

Reply:
@jpadilla I fixed the type definitions. I'm glad if you check and merge it.